### PR TITLE
dbx-deploy paths fix & repo force update

### PR DESCRIPTION
**Pull request by StanislavZoldak**
* `dbxdeploy.target.repo.root_dir` and `dbxdeploy.target.repo.path` deprecated in `dbxdeploy.target.repo.base_dir`
* `dbx:repo:update` command now supports the `--force` argument to force update of a repo in Databricks repo (similar to `git reset --hard`)

### CODE REVIEW

File: src/dbxdeploy/repos/RepoManager.py [hunk 1]
```diff
@@ -10,22 +11,35 @@ def __init__(
...
         self.__repos_api = repos_api
         self.__workspace_api = workspace_api
-        self.__logger = logger
-        self.__repo_root_dir = repo_root_dir
-        self.__repo_path = repo_path
 
-    def create_or_update(self, repo_url, branch, tag, repo_name, force):
-        repo_path = self.__repo_path.format(repo_name=repo_name)
+    def create_or_update(self, repo_url, repo_name=None, branch=None, tag=None, force=False):
+        if repo_name is None and "{repo_name}" in self.__repo_base_dir:
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:30]: Please use some common validation function like validate(repo_name, "{repo_name}", "--repo_name")

File: src/dbxdeploy/repos/RepoManager.py [hunk 2]
```diff
@@ -48,12 +62,18 @@ def create_or_update(self, repo_url, branch, tag, repo_name, force):
...
             raise e
+        else:
+            self.__logger.info("Repo successfully updated")
 
-        self.__logger.info("Repo successfully updated")
-
-    def delete(self, repo_url, repo_name):
-        repo_path = self.__repo_path.format(repo_name=repo_name)
+    def delete(self, repo_url, repo_name, branch, tag):
+        if repo_name is None and "{repo_name}" in self.__repo_base_dir:
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:69]: Why do you duplicate this logic?

File: src/dbxdeploy/repos/RepoManager.py [hunk 3]
```diff
@@ -48,12 +62,18 @@ def create_or_update(self, repo_url, branch, tag, repo_name, force):
...
+            raise Exception("{repo_name} used in repo.base_dir, but --repo_name argument was not provided")
+        if branch is None and "{branch_name}" in self.__repo_base_dir:
+            raise Exception("{branch_name} used in repo.base_dir, but --branch argument was not provided")
+        if tag is None and "{tag_name}" in self.__repo_base_dir:
+            raise Exception("{tag_name} used in repo.base_dir, but --tag argument was not provided")
 
+        repo_path = self.__repo_base_dir.format(repo_name=repo_name, branch_name=branch, tag_name=tag)
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:76]: Let's create __resolve_repo_path() method which handles validation and format self.__repo_base_dir

**Review by kutny — CHANGES_REQUESTED** [review]

File: src/dbxdeploy/repos/RepoUpdateCommand.py [hunk 4]
```diff
@@ -15,25 +15,22 @@ def get_description(self):
...
     def run(self, input_args: Namespace):
-        if not (input_args.repo_url and input_args.repo_name) or not (input_args.branch or input_args.tag):
+        if not input_args.repo_url:
             raise Exception("Missed required arguments. Check -h for the list of required arguments.")
-        if input_args.branch and input_args.tag:
-            raise Exception("Can't pick both --checkout-branch and --checkout-tag. Check -h for details.")
 
         self.__repo_manager.create_or_update(
             input_args.repo_url,
+            input_args.repo_name,
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoUpdateCommand.py:32]: Let's pass all optional arguments as dict

File: src/dbxdeploy/repos/RepoDeleteCommand.py [hunk 5]
```diff
@@ -15,11 +15,14 @@ def get_description(self):
 
     def configure(self, argument_parser: ArgumentParser):
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoDeleteCommand.py:16]: Most of the logic is duplicated in RepoDeleteCommand x RepoUpdateCommand

**Review by kutny — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 6]
```diff
@@ -43,6 +43,7 @@ flake8-fixme = "^1.1.1"
 pep8-naming = "^0.11.1"
 poethepoet = "^0.10.0"
 black = "^20.8b1"
+daipe-core = "1.0.0"
```
Comment (kutny) [inline pyproject.toml:46]: `>=1.0`

**Review by kutny — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 6 above)
Comment (kutny) [reply pyproject.toml:46]: Better to put daipe-core to normal dependencies (non-dev) like in https://github.com/daipe-ai/databricks-bundle/blob/master/pyproject.toml#L34

File: src/dbxdeploy/repos/RepoManager.py [hunk 7]
```diff
@@ -10,22 +11,28 @@ def __init__(
...
+        repo_group_path = os.path.split(repo_path)[0]
         self.__logger.info(f"Updating repo at {repo_path}")
-        self.__workspace_api.mkdirs(self.__repo_root_dir)
+        self.__workspace_api.mkdirs(repo_group_path)
 
-        repo_list = self.__repos_api.list(self.__repo_root_dir, None).get("repos", [])
+        repo_list = self.__repos_api.list(repo_group_path, None).get("repos", [])
 
         try:
             repo_id = self.__repos_api.get_repo_id(repo_path)
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:38]: Place self.__check_duplicates(repo_list) before try

File: src/dbxdeploy/repos/RepoManager.py [hunk 8]
```diff
@@ -77,10 +93,13 @@ def __check_duplicates(self, repo_list: list):
         if len(duplicated_branches) > 0:
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:93]: __check_duplicates() should also check tags, not only branches

File: src/dbxdeploy/repos/RepoManager.py [hunk 9]
```diff
@@ -77,10 +93,13 @@ def __check_duplicates(self, repo_list: list):
...
             raise Exception(f"Following branches are duplicated in current env: {', '.join(duplicated_branches)}.")
 
-    def __check_valid_branch(self, repo_list: list, branch: str):
-        repo_branches = [repo.get("branch") for repo in repo_list]
-        if branch in repo_branches:
-            raise Exception(f"Branch {branch} is already present in current env.")
+    def __check_valid_branch(self, repo_list: list, branch: str, tag: str):
```
Comment (kutny) [inline src/dbxdeploy/repos/RepoManager.py:96]: Remove this completely

**Review by kutny — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 10]
```diff
@@ -35,6 +35,7 @@ pyfony-core = "^0.8.1"
 console-bundle = "^0.5"
 Jinja2 = "^2.0.0"
 databricks-cli = "^0.16.0"
+daipe-core = "<=1.0.0"
```
Comment (kutny) [inline pyproject.toml:38]: Fix

**Review by kutny — CHANGES_REQUESTED** [review]

File: src/dbxdeploy/repos/RepoUpdateCommand.py (hunk 4 above)
Comment (StanislavZoldak) [reply src/dbxdeploy/repos/RepoUpdateCommand.py:32]: Agreed to not do this right now

File: src/dbxdeploy/repos/RepoManager.py (hunk 8 above)
Comment (StanislavZoldak) [reply src/dbxdeploy/repos/RepoManager.py:93]: Tag key not supperted in list-repo Repos Api. Will revert to branches only.

**Review by kutny — APPROVED** [review]

**Pull request merged**